feat: adjusted Java and MongoDB connection for first running

vitormmartins opened the pull request:
Summary
The Pull Request introduces a major architectural overhaul of the ChatForge application by refactoring the codebase to adopt Hexagonal Architecture (Ports and Adapters) alongside Domain-Driven Design (DDD) principles. Additionally, the PR implements new features, enriches the build and deployment process, and enhances the test structure.
Main Changes
Architecture Refactor:
Transitioned to Hexagonal Architecture with clearly defined Domain, Application, Infrastructure, and Web layers.
Emphasized separation of concerns with proper usage of domain services, value objects, and repository interfaces.
Security Updates:
Implemented JWT and cookie-based authentication via custom Spring Security configurations and filters.
Added adapters for encoding passwords and managing user authentication.
Infrastructure Enhancements:
Created a Dockerfile and docker-compose.yml for streamlined application setup and deployment.
Integrated Flyway for managing database migrations.
API Improvements:
Introduced OpenAPI specifications for generating API documentation.
Refactored controllers to interact with application services rather than repositories directly.
Testing Improvements:
Added comprehensive unit and integration tests for key components and configurations.
Utilized mock classes and in-memory databases for isolated test scenarios.
Miscellaneous:
Added new utility classes for security (e.g., JwtUtil) and domain validation (e.g., Username, Email value objects).
Created migration and architectural documentation for smoother transitions to the new structure.

Commit:
feat: adjusted Java and MongoDB connection for first running

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/controller/AuthController.java:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  @PostMapping("/login")
+  public ResponseEntity<?> login(@RequestBody User user) {
```
Definir um "model" (DTO, Data Transfer Object) para o endpoint.
O modelo de "User" também define o "Id", chave do objeto, e não deve ser esperado que o front envie isso durante o login

vitormmartins replied:
Feito

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/controller/AuthController.java:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  @PostMapping("/register")
+  public ResponseEntity<?> register(@RequestBody User user) {
```
Mesmo caso do login. Melhor um model separado para o endpoint com todos os dados que são esperados do front.

vitormmartins replied:
Feito

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/filter/JwtFilter.java:
```diff
@@ -0,0 +1,52 @@
+    String token = authHeader.substring(7);
+    String username = jwtUtil.extractUsername(token);
+
+    if (username != null && SecurityContextHolder.getContext().getAuthentication() == null) {
```
Revisar essa sequência. Está confusa e tem algumas funções que estão sendo chamadas repetidamente, como por exemplo "extractUsername"

vitormmartins replied:
Feito

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/util/JwtUtil.java:
```diff
@@ -0,0 +1,53 @@
+@Component
+public class JwtUtil {
+
+  private static final String SECRET_KEY = "your-256-bit-secret-your-256-bit-secret"; // Change this!
```
Se for fazer essa aplicação escalar horizontalmente, esse parametro tem que ser passado externamente para a aplicação, pois tem que ser a mesma em todas as instâncias.

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/util/JwtUtil.java:
```diff
@@ -0,0 +1,53 @@
+            .getBody();
+  }
+
+  public boolean validateToken(String token, String username) {
```
a nomenclatura desse método não corresponde com o abaixo, que também retorna um booleano.
eu, pessoalmente, prefiro a nomenclatura com o "is" na frente para ficar escrito como se fosse uma pergunta.

Não deveria ser necessário passar o nome de usuário e nem verificar se o token está expirado. somente verificar a assinatura é mais que suficiente.

Commit:
fix: adjusted JWT claims on filter

Commit:
fix: adjusted @Value initializer for JwtUtil

Commit:
fix: fixed /login endpoint for basic JWT generation

Commit:
fix: generating JWT after name and password authentication

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/filter/JwtFilter.java:
```diff
@@ -0,0 +1,49 @@
+                                  @NonNull HttpServletResponse response,
+                                  @NonNull FilterChain chain)
+          throws ServletException, IOException {
+    if(request.getRequestURI().equals("/auth/login") || request.getRequestURI().equals("/auth/register")) {
```
Ignorar esses dois endpoints, que não requerem autenticação, como parte de configuração do spring e não como uma condicional hardcoded

vitormmartins replied:
if's foram retirados do filtro

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/filter/JwtFilter.java:
```diff
@@ -0,0 +1,49 @@
+      chain.doFilter(request, response);
+      return;
+    }
+    String authHeader = request.getHeader("Authorization");
```
possível melhoria é dar suporte para autenticação por cookie.
consultar como que cookies funcionam e qual é a melhor configuração para um cookie de autenticação.

vitormmartins replied:
Autenticação por cookie feita com filtro e cookie login

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/service/CustomAuthenticationProvider.java:
```diff
@@ -0,0 +1,37 @@
+
+    UserDetails userDetails = customUserDetailsService.loadUserByUsername(name);
+
+    if (!userDetails.getUsername().equals(name) || !userDetails.getPassword().equals(password)) {
```
não precisa implementar para esse PR, mas implementar salting e hashing de senhas de usuário.

renatommartins replied:
referencias: Argon2 e random criptografico

renatommartins replied:
https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html

vitormmartins replied:
Usar método de gerador random criptográfico em Java já pronto (pesquisar).

vitormmartins replied:
Feito usando a classe do Spring Security PasswordEncoder

renatommartins commented on src/test/java/io/vitormmartins/chatforge/spring_chatforge/filter/JwtFilterTest.java:
```diff
@@ -0,0 +1,29 @@
+import org.mockito.MockitoAnnotations;
+import org.springframework.security.core.userdetails.UserDetailsService;
+
+public class JwtFilterTest {
```
Se for implementar testes unitários, seria uma boa já completar os testes unitários do que foi feito nesse PR.

vitormmartins replied:
Feito para a classe JwtFilter

vitormmartins replied:
Testes de autenticação feitos

renatommartins commented on src/main/resources/application.properties:
```diff
@@ -1 +1,16 @@
+spring.datasource.hikari.idleTimeout=600000
+spring.datasource.hikari.maxLifetime=1800000
+
+jwt.secret-key=OBslwXQbeJPMIq2bNCmAADfBF36Tda8BhRrXebM8zG2P2ksGSCol9f9bDZpVH6gn
```
Existe alguma forma de sobrescrever/adicionar arquivos de configuração por ambiente? Commitar chaves secretas em um repositório é má pratica. Se for possível o que eu citei, a chave secreta ficaria em um arquivo não rastreado na maquina e seria configurado no deploy de ambiente.

vitormmartins replied:
Será feito quando for deployado em cloud

renatommartins commented on src/main/java/io/vitormmartins/chatforge/spring_chatforge/service/CustomAuthenticationProvider.java:
```diff
@@ -0,0 +1,37 @@
+  private final CustomUserDetailsService customUserDetailsService;
+
+  @Override
+  public Authentication authenticate(final Authentication authentication) throws AuthenticationException {
```
implementação de migration para o banco de dados

vitormmartins replied:
Implementação de migration para o BD terminada.

Commit:
feat: included JwtFilter unit tests.

Commit:
feat: removed JwtAuthenticationFilter if's to ignore /auth/login and /auth/register endpoints

Commit:
feat: add cookie-based authentication endpoint and filter and included Postman Collection for project as documentation

Commit:
Add password encoding and validation for authentication

Integrated `PasswordEncoder` to securely hash passwords during user registration and validate them during login. Updated `AuthController`, `CustomAuthenticationProvider`, and `SecurityConfig` to support these changes. Adjusted Postman collection tests to reflect updated data and endpoints.

Commit:
feat: add Flyway migration support and update User entity with timestamps

Commit:
Refactor JWT handling and improve authentication flow

Replaced redundant filter implementations with an abstract base class (`AbstractJwtFilter`) for shared logic. Introduced custom exceptions (`InvalidTokenException` and `TokenExpiredException`) for better error handling. Updated tests and controller methods to reflect refinements in token processing and response patterns.

Commit:
Add OpenAPI support and improve authentication flow

Integrated OpenAPI with Springdoc for API documentation, including Swagger UI. Updated security configuration to allow access to API docs. Fixed logic in authentication filters to correctly handle token processing outcomes. Removed redundant database creation in the migration script.

renatommartins commented on postman/collections/ChatForge.postman_collection.json:
```diff
@@ -0,0 +1,733 @@
+			]
+		},
+		{
+			"name": "Get User Cookie auth",
```
eu só recebi 401 nesse endpoint mesmo enviando o cookie de autenticação

vitormmartins replied:
Testar agora com a nova Collection do Postman, gerando o token e depois testando com o GET.

renatommartins commented on postman/collections/ChatForge.postman_collection.json:
```diff
@@ -0,0 +1,733 @@
+					}
+				],
+				"url": {
+					"raw": "http://localhost:8080/get/user/V%C3%ADtor%20Matosinho",
```
não está usando a variavel de ambiente para o path

vitormmartins replied:
Testar agora com a nova Collection.

renatommartins commented on postman/collections/ChatForge.postman_collection.json:
```diff
@@ -0,0 +1,733 @@
+					}
+				},
+				"url": {
+					"raw": "http://localhost:8080/auth/register",
```
não está usando a variavel do ambiente

vitormmartins replied:
Testar agora com a Collection atualizada

renatommartins commented on src/test/java/io/vitormmartins/chatforge/spring_chatforge/filter/JwtFilterTest.java:
```diff
@@ -0,0 +1,189 @@
+// File: src/test/java/io/vitormmartins/chatforge/spring_chatforge/filter/JwtFilterTest.java
```
Testes precisam mockar o db

renatommartins commented on HELP.md:
```diff
@@ -0,0 +1,154 @@
+# Read Me First
```
adicionar comandos de build e execução da aplicação

renatommartins commented on HELP.md:
````diff
@@ -0,0 +1,154 @@
+To create a Docker container for MySQL, use the following command:
+
+```shell
+sudo docker run --name some-mysql -e MYSQL_ROOT_PASSWORD=my-secret-pw -p 3306:3306 -d mysql:8.3
````
criar o db junto com o container
sudo docker run --name some-mysql -e MYSQL_ROOT_PASSWORD=my-secret-pw -e MYSQL_DATABASE=mysqldatabase -p 3306:3306 -d mysql:8.3

uma ideia é adicionar script docker compose para subir o ambiente local. Também sugiro fazer um segundo script docker compose para rodar o ambiente todo (como se fosse um deploy)

vitormmartins replied:
Primeiro script Docker Compose adicionado. Neste momento, já está subindo o ambiente todo pelo que entendi. É necessário um segundo script para subir o ambiente todo ainda?

vitormmartins replied:
O banco já está sendo criado no script.

vitormmartins committed:
feat: implement user authentication and registration DTOs, commands, and services using Domain Driven Design

renatommartins commented on src/main/java/io/vitormmartins/chatforge/application/user/dto/RegisterUserCommand.java:
```diff
@@ -0,0 +1,19 @@
+ */
+public record RegisterUserCommand(
+    String username,
+    String email,
```
email é obrigatório? se não for, é uma boa deixar um comentário explicitando a intenção ou colocar atributo com nullable

renatommartins commented on src/main/java/io/vitormmartins/chatforge/application/user/service/UserApplicationService.java:
Classe de autenticação de usuário duplicada. uma classe com essa funcionalidade já existia

vitormmartins replied:
Acredito que a duplicação não existe mais. Pode passar a classe que estava com essa funcionalidade duplicada?

Commit:
feat: implement user registration and authentication improvements with JWT and cookie support

Commit:
feat: fixed circular dependency on SecurityConfig and DomainAuthenticationProvider, update on application configuration, and add Docker support for improved deployment.

Commit:
feat: update application properties and pom.xml for improved configuration and dependency management

Commit:
fix: update SecurityConfig to permit access to actuator endpoints and adjust swagger UI permissions

Commit:
feat: enhance JWT authentication with improved error handling and add in-memory database support for testing

Commit:
feat: update Postman collection and environment for improved API testing and update endpoints for user management

Commit:
feat: update API versioning in Auth and User controllers, enhance Postman collection for improved testing
fix: fixed cookie token creation

Commit:
feat: implement cookie-based JWT authentication support in filters

Commit:
fix: refactor JWT authentication filter to improve token processing and update Postman collection for dynamic URL usage

Commit:
docs: update HELP.md to enhance application running instructions and Docker setup guidance

Commit:
test: add integration tests for authentication and user management endpoints

Commit:
```
WIP - Add test configurations and unit tests for AuthController and UserController

- Added application-test.properties for testing configurations
- Implemented AuthControllerTest with mocked dependencies
- Added UserControllerTest to test user-related functionalities
- Modified ChatForgeApplication.java to include MongoDataAutoConfiguration
- Updated ChatForgeApplicationTests to import necessary user service and DTOs
```

Commit:
```
Added test configuration and unit tests for authentication and user controller

- Added application-test.properties to configure testing environment
- Created AuthControllerTest for testing authentication logic
- Added UserControllerTest to validate user retrieval functionality
- Modified ChatForgeApplication and ChatForgeApplicationTests to include necessary imports
```

renatommartins commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM maven:3.8.6-eclipse-temurin-17 AS build
+WORKDIR /app
+COPY pom.xml .
+COPY src ./src
```
faltando comando de copy do spec do openapi

vitormmartins replied:
Feito: openapi ./openapi

renatommartins commented on HELP.md:
```diff
@@ -0,0 +1,264 @@
+cd /path/to/spring-chatforge
+
+# Build the application with Maven
+./mvnw clean package
```
comando -DskipTests causa erros ao buildar o código de teste

vitormmartins replied:
Corrigido. Na verdade, -DskipTests pula os testes. Era erro de falta de dois métodos, PUT e DELETE, que tinham na interface mas não estavam no Controller.

renatommartins commented on src/main/java/io/vitormmartins/chatforge/config/SecurityConfig.java:
```diff
@@ -0,0 +1,75 @@
+                .requestMatchers("/swagger-ui/**").permitAll()
+                .requestMatchers("/swagger-ui.html").permitAll()
+                .anyRequest().authenticated())
+            .exceptionHandling(exception -> exception
```
Está capturando erro de offset date time

vitormmartins replied:
Corrigido. Ao testar o POST login, limpar os Cookies da request antes de envia-la no Postman.

renatommartins commented on HELP.md:
````diff
@@ -0,0 +1,264 @@
+
+```shell
+# Run the JAR file
+java -jar target/spring-chatforge-0.0.1-SNAPSHOT.jar
````
alertar para definir a senha do db como variavel de ambiente

vitormmartins replied:
Feito no HELP.md. Está em uma observação antes de buildar a aplicação.

renatommartins commented on postman/collections/ChatForge.postman_collection.json:
atualizar coleção do postman. Endpoints estão sem o prefixo "v1"

vitormmartins replied:
Corrigido, checar nova collection.

Commit:
refactor: migrate domain models to Java Records, update service and repository methods accordingly

- Simplified `Email`, `UserId`, `Password`, and `Username` domain models by converting them to Java Records.
- Refactored code to utilize direct field access (`value()`) instead of getter methods.
- Updated JPA mapper methods and repositories to align with domain model changes.
- Enhanced OpenAPI specifications with improved operationId definitions for better documentation.
- Added new authentication endpoints (`update-password` and `delete-account`) in controllers and API specifications (with placeholders for implementation).
- Included missing `openapi` directory in `Dockerfile`.

Commit:
feat: implement JWT authentication and update user service to use records

Commit:
test: add unit tests for user-related commands and services

- Added unit tests for `RegisterUserCommand` and `AuthenticateUserCommand`.
- Implemented tests for `UserApplicationService` covering registration, authentication, and username availability checks.
- Added `PasswordEncoderPortTest` to validate password encoding contract.

renatommartins committed:
sugestões de código

Commit:
test: add unit tests for ApplicationConfig and PasswordConfig, and refactor UserDomainService to use records

Commit:
Merge remote-tracking branch 'origin/1-implementar-autenticação-jwt-suggestion-250824' into 1-implementar-autenticação-jwt

# Conflicts:
#	src/main/java/io/vitormmartins/chatforge/domain/user/service/UserDomainService.java

Commit:
refactor: replace domain-specific models with standard types and improve test consistency

- Replaced `UserId` and `Username` with `long` and `String` respectively across codebase.
- Updated session and token processing logic in `JwtAuthenticationFilter` for better handling.
- Refactored existing tests (`ApplicationConfigTest`, `UserApplicationServiceTest`, `AuthControllerTest`) to align with new model changes and improve readability.
- Simplified mocked object creation and dependency injection in tests.
